File: data_structures/queue/ruby/queue.rb
class Queue
    def initialize
        @store = Array.new
    end

    def enqueue(data)
        @store.push(data)
    end

    def dequeue()
        raise Exception if size == 0
        @store.shift
    end

    def front()
        raise Exception if size == 0
        @store.first
    end

    def back()
        @store.last
    end

    def size()
        @store.size
    end
end

# Usage
Q = Queue.new
Q.enqueue(1)
Q.enqueue(2)
Q.enqueue(3)

puts Q.front()
puts Q.back()

puts Q.dequeue()
puts Q.dequeue()

File: data_structures/queue/js/queue.js
class Queue {
  constructor () {
    this.store = []
  }
  enqueue (data) {
    this.store.push(data)
  }
  dequeue () {
    if (this.store.length === 0)
      throw 'Zero length'
    return this.store.shift()
  }
  front () {
    if (this.store.length === 0)
      throw 'Zero length'
    return this.store[0]
  }
  back () {
    return this.store[this.store.length - 1]
  }
  size () {
    return this.store.length
  }
}

// usage
const q = new Queue()
q.enqueue('a')
q.enqueue('b')
q.enqueue('c')

console.log('-----')
console.log(q.front())
console.log(q.back())
console.log('-----')
console.log(q.dequeue())
console.log(q.dequeue())
